File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        n = sum(nums)
        if n%2!=0:
            return False
        ss = n//2
        dp = [False]*(ss+1)
        dp[0] = True
        for num in nums:
            for i in range(ss,num-1,-1):
                dp[i] = dp[i] or dp[i-num]
        return dp[ss]

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int num:nums){
            sum += num; 
        }
        if(sum%2!=0){
            return false;
        }
        int ss = sum/2;
        boolean[] dp = new boolean[ss+1];
        dp[0] = true;
        for(int num:nums){
            for(int i=ss;i>num-1;i--){
                dp[i] = dp[i] || dp[i-num];
            }
        }
        return dp[ss];
    }
}